src/artemis/fgs_communicator.py

```diff
@@ -0,0 +1,144 @@
+
+import artemis.log
+from artemis.ispyb.store_in_ispyb import StoreInIspyb2D, StoreInIspyb3D
+from artemis.nexus_writing.write_nexus import (
+    NexusWriter,
+    create_parameters_for_first_file,
+    create_parameters_for_second_file,
+)
+from artemis.parameters import FullParameters
+from artemis.zocalo_interaction import run_end, run_start, wait_for_result
+
+
+class FGSCommunicator(CallbackBase):
+    """Class for external communication (e.g. ispyb, zocalo...) during Artemis
+    grid scan experiments.
+
+    Listens to documents emitted by the RE and:
+    - prepares nexus files
+    - prepares ipsyb deposition
+    - submits job to zocalo
+    """
+
+    def __init__(self):
+        self.reset(FullParameters())
+
+    def reset(self, parameters: FullParameters):
+        self.params = parameters
+        ispyb_config = os.environ.get("ISPYB_CONFIG_PATH", "TEST_CONFIG")
+        self.ispyb = (
+            StoreInIspyb3D(ispyb_config, self.params)
+            if self.params.grid_scan_params.is_3d_grid_scan
+            else StoreInIspyb2D(ispyb_config, self.params)
+        )
+        self.active_uid = None
+        self.gridscan_uid = None
+        # TODO add flag for this or delete?
+        # useful for testing, to not duplicate files:
+        # self.params.detector_params.prefix += str(time.time())
+        self.results = None
+        self.processing_time = 0.0
+        self.nxs_writer_1 = NexusWriter(create_parameters_for_first_file(self.params))
+        self.nxs_writer_2 = NexusWriter(create_parameters_for_second_file(self.params))
+        self.datacollection_group_id = None
+        self.xray_centre_motor_position = None
+        self.ispyb_ids: tuple = (None, None, None)
+        self.descriptors: dict = {}
+
+    def start(self, doc: dict):
+        artemis.log.LOGGER.debug(f"\n\nReceived start document:\n\n {doc}\n")
+        self.active_uid = doc.get("uid")
+        if doc.get("plan_name") != "run_gridscan_and_move":
+            return
+        self.gridscan_uid = doc.get("uid")
+
+        artemis.log.LOGGER.info(f"Creating Nexus files for run {self.active_uid}")
+        self.nxs_writer_1.create_nexus_file()
+        self.nxs_writer_2.create_nexus_file()
+
+        artemis.log.LOGGER.info(f"Initialising Zocalo for run {self.active_uid}")
+
+    def get_descriptor_doc(self, key) -> dict:
+        associated_descriptor_doc = self.descriptors.get(key)
+        if type(associated_descriptor_doc) is not dict:
+            raise ValueError(
+                "Non-dict object stored in descriptor list, or key not valid."
+            )
+        return associated_descriptor_doc
+
+    # TODO is this going to eat too much memory if there are a lot of these with a lot of data?
+    # It probably gets destroyed on self.reset() ?
+    def descriptor(self, doc: dict):
+        artemis.log.LOGGER.debug(f"\n\nReceived descriptor document: {doc}\n")
+        self.descriptors[doc.get("uid")] = doc
```

Comment:
As discussed, we may not need this anymore if we no longer need a reference to the plan. Let's see if it falls out in refactoring

Reply:
yes, it's much cleaner now fgs_communicator doesn't have to know which plan it's attached to